### ISSUE TITLE
Design MockitoSession API improvements for unit test frameworks

### ISSUE THREAD

**Issue by szczepiq** [mockito/mockito#898]
It is desired that JUnit rule / runner uses MockitoSession API (#865) so that they dogfood our public API. In order to do that we need to do more design work.

Mockito Session API (#865) is geared towards using it in 'setup' and 'tear down' methods of the unit test framework. In 'tear down' method, the user does not have access to the exception (test failure). In JUnit rule, we do have access to the exception and we can leverage it for better user experience. We can prevent dual failure when validating mockito usage fails, on top of the actual test failure. Dual failure could be confusing to the user. When Mockito Session API is used, we cannot really avoid dual failure. Dual failure is relatively minor problem, the test failure is reported by JUnit first, before any potential framework validation problem.

JUnit runner is even more interesting. Unlike the rule, the runner has access to extra test lifecycle events like "before all tests" and "after all tests".

This ticket is a placeholder for design work on growing MockitoSession API so that it can be used by runner and the rule. While growing MockitoSession API we can potentially expose MockitoTestListener. All that work has lower priority than getting the MockitoSession API released. I suggest we even wait for user's feedback before we start growing MockitoSession API.

**Comment by TimvdLippe** [comment]
This was fixed in #865
